transcode/transcode.go

```diff
@@ -115,59 +115,40 @@ func RunTranscodeProcess(transcodeRequest TranscodeSegmentRequest, streamName st
-			// Get renditions from remote broadcaster
-			tr, err = broadcasterClient.TranscodeSegmentWithRemoteBroadcaster(rc, int64(segmentIndex), transcodeProfiles, streamName, u.DurationMillis)
-			if err != nil {
-				return outputs, fmt.Errorf("failed to run TranscodeSegmentWithRemoteBroadcaster: %s", err)
-			}
-		} else {
-			// Get renditions from local broadcaster
-			tr, err = localBroadcasterClient.TranscodeSegment(rc, int64(segmentIndex), transcodeProfiles, u.DurationMillis, manifestID)
-			if err != nil {
-				return outputs, fmt.Errorf("failed to run TranscodeSegment: %s", err)
-			}
-		}
-		// Store renditions
-		for _, transcodedSegment := range tr.Renditions {
-			renditionIndex := getProfileIndex(transcodeProfiles, transcodedSegment.Name)
-			if renditionIndex == -1 {
-				return outputs, fmt.Errorf("failed to find profile with name %q while parsing rendition segment", transcodedSegment.Name)
-			}
-
-			relativeRenditionPath := fmt.Sprintf("rendition-%d/", renditionIndex)
-			relativeRenditionURL, err := url.Parse(relativeRenditionPath)
-			if err != nil {
-				return outputs, fmt.Errorf("error building rendition segment URL %q: %s", relativeRenditionPath, err)
-			}
-			renditionURL := targetOSURL.ResolveReference(relativeRenditionURL)
-
-			err = clients.UploadToOSURL(renditionURL.String(), fmt.Sprintf("%d.ts", segmentIndex), bytes.NewReader(transcodedSegment.MediaData))
-			if err != nil {
-				return outputs, fmt.Errorf("failed to upload master playlist: %s", err)
+		queue <- segmentInfo{Input: u, Index: segmentIndex}
+	}
+	close(queue)
+	// Use channel for recieving the errors
+	errors := make(chan error, 100)
+	// Start number of workers in parallel
+	var completed sync.WaitGroup
+	completed.Add(config.TranscodingParallelJobs)
+	for index := 0; index < config.TranscodingParallelJobs; index++ {
+		go func() {
+			defer completed.Done()
+			for segment := range queue {
+				err := transcodeSegment(segment, streamName, manifestID, transcodeRequest, transcodeProfiles, targetOSURL)
+				if err != nil {
+					errors <- err
+					return
+				}
+				var completedRatio = calculateCompletedRatio(len(sourceSegmentURLs), segment.Index+1)
+				if err = clients.DefaultCallbackClient.SendTranscodeStatus(transcodeRequest.CallbackURL, clients.TranscodeStatusTranscoding, completedRatio); err != nil {
+					log.LogError(transcodeRequest.RequestID, "failed to send transcode status callback", err, "url", transcodeRequest.CallbackURL)
+				}
 			}
-		}
-
-		var completedRatio = calculateCompletedRatio(len(sourceSegmentURLs), segmentIndex+1)
-		if err = clients.DefaultCallbackClient.SendTranscodeStatus(callbackURL, clients.TranscodeStatusTranscoding, completedRatio); err != nil {
-			log.LogError(transcodeRequest.RequestID, "failed to send transcode status callback", err, "url", callbackURL)
-		}
+		}()
+		// Add some desync interval to avoid load spikes on segment-encode-end
+		time.Sleep(713 * time.Millisecond)
```

Comment:
Feels pretty arbitrary - are the load spikes on GPU or CPU? Did you see any issues with the load esp if it's GPU?